languages/bash.scm

```diff
@@ -23,13 +32,88 @@
   (if_statement)
   (list)
   (pipeline)
+  (redirected_statement)
   (subshell)
   (variable_assignment)
   (while_statement)
-  ; TODO: etc.
 ] @allow_blank_line_before
 
-; Surround with spaces
+; Insert a new line after multi-line syntactic blocks or, for where
+; single-line variants exists, after the "closing" subnodes (the
+; specificity is to avoid targeting the single-line context)
+[
+  (if_statement)
+  (case_statement)
+  (do_group)
+] @append_hardline
+
+(subshell
+  ")" @append_empty_softline
+  .
+)
+
+(compound_statement
+  "}" @append_empty_softline
+  .
+)
+
+; A run of "units of execution" (see below, sans variables which are
+; special) and function definitions should be followed by a new line,
+; before a multi-line syntactic block or variable.
+(
+  [
+    (command)
+    (compound_statement)
+    (function_definition)
+    (list)
+    (pipeline)
+    (redirected_statement)
+    (subshell)
+  ] @append_empty_softline
+  .
+  [
+    (c_style_for_statement)
+    (case_statement)
+    (comment)
+    (compound_statement)
+    (declaration_command)
+    (for_statement)
+    (function_definition)
+    (if_statement)
+    (subshell)
+    (variable_assignment)
+    (while_statement)
+  ]
```

Comment:
What would go wrong if you simply dropped this part?

Reply:
(Same as https://github.com/tweag/topiary/pull/182#discussion_r1083931668: Doing so puts blanks lines where they're unnecessary, including places where they're syntactically invalid.)